BUG: Fixed Von Mises distribution for big values of kappa

bashtage opened the pull request:
Extends branch in #17378 to prevent fix applying to RandomState and changing its stream.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

Androp0v committed:
Fixed Von Mises distribution for big values of kappa by falling back to a normal distribution (which the von mises distribution converges to).

Androp0v committed:
Fixed style and added check for bounds in [-pi,pi] interval.

Androp0v committed:
Added test for [-np.pi, np.pi] range for the von mises distribution using large values of kappa (>1e6)

bashtage committed:
BUG: Prevent RandomState from changing

Apply vonmises fix only to Generator
Add tests for correctness

closes #17378
closes #17275

charris commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -853,9 +854,30 @@ double random_vonmises(bitgen_t *bitgen_state, double mu, double kappa) {
+          return result - 2*M_PI;
+        }
+        return result;
+      } else {
```
Indentation needs to be 4 spaces, `else` on new line.

charris replied:
Style needs fixing all round, let's leave that to a later cleanup.

bashtage replied:
While I don't mind changing the commit, the entire file is formatted using these conventions (2 spaces, single like else).  IIRC think this is Google C format as formatted using clang-format.

charris commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -853,9 +854,30 @@ double random_vonmises(bitgen_t *bitgen_state, double mu, double kappa) {
-      double rho = (r - sqrt(2 * r)) / (2 * kappa);
-      s = (1 + rho * rho) / (2 * rho);
+#ifndef NP_RANDOM_LEGACY
+      /* Fallback to normal distribution for big values of kappa
```
First comment line should be blank.

rkern commented:
I think that it would be safer/more readable/more convenient to do the `#ifdef NP_RANDOM_LEGACY` switch between two copies of the whole function. Then we have free reign in the non-legacy copy of the function. Since we're not maintaining the legacy version, there's no DRY benefit to sharing the (thus-far) common parts of the function. I think it would make the task of verifying that the legacy variant remains unchanged much, much simpler. And it would be an easy convention that new contributors can follow, so we don't have to police them with followup PRs like this.

charris commented:
The code coverage failure should probably be fixed as it seems a common case.

bashtage committed:
CLN: Move to legacy function

Avoid conditional compilation and move old version to legacy_vonmises
Small clean up
Additional comments

bashtage commented:
> I think that it would be safer/more readable/more convenient to do the `#ifdef NP_RANDOM_LEGACY` switch between two copies of the whole function. Then we have free reign in the non-legacy copy of the function. Since we're not maintaining the legacy version, there's no DRY benefit to sharing the (thus-far) common parts of the function. I think it would make the task of verifying that the legacy variant remains unchanged much, much simpler. And it would be an easy convention that new contributors can follow, so we don't have to police them with followup PRs like this.

I moved it to legacy-distributions.c as legacy_vonmises.  This has the advantage of clarifying that these functions should not be altered except under special circumstances.

bashtage commented on numpy/random/tests/test_generator_mt19937.py:
```diff
@@ -1735,6 +1735,26 @@ def test_vonmises_nan(self):
+        else:
+            assert np.all(random_state_vals != gen_vals)
+
+    @pytest.mark.parametrize("mu", [-7., -np.pi, -3.1, np.pi, 3.2])
+    @pytest.mark.parametrize("kappa", [1e-9, 1e-6, 1, 1e3, 1e15])
+    def test_vonmises_large_kappa_range(self, mu, kappa):
+        r = random.vonmises(mu, kappa, 50)
+        assert_(np.all(r > -np.pi) and np.all(r <= np.pi))
+
```
I think that coverage is a bit suspicious.  I added this test to ensure that all code paths are hit.

rkern commented:
> I moved it to legacy-distributions.c as legacy_vonmises.

Agreed.

charris commented:
The azure failure is unrelated, I didn't rerun the test as the output would be lost.

charris commented:
Thanks @bashtage .
